src/ApnPrebidVast.js

```diff
@@ -0,0 +1,350 @@
+	    	callback(false);
+	    };
+	    molScr.async = true;
+	    molScr.type = 'text/javascript';
+	    molScr.src = '//acdn.adnxs.com/video/plugins/mol/videojs_5.vast.vpaid.min.js';
+	    var node = document.getElementsByTagName('head')[0];
+	    node.appendChild(molScr);
+	}
+	else {
+    	_logger.log(_prefix, 'Appnexus MailOnline Plugin already loaded');
+		callback(true);
+	}
+}
+
+(function () {
+	// if bidders settings are present in the window.apn_plugin_prebid_options variable load prebid.js and do the bidding
+	if (window.apn_plugin_prebid_options && window.apn_plugin_prebid_options.biddersSpec) {
+		APN_prebid_in_progress = true;
+		loadPrebidScript(window.apn_plugin_prebid_options, true);
+	}
+	loadApnMolPlugin(function() {});
+})();
+
+// register videojs prebid plugins
+function registerPrebidVastPlugin(vjs) {
+	_vjs = vjs;
+	if (!_vjs.getPlugins().apnPrebidVastPlugin) {
+		reqisterPrebidVastPlugin();
+	}
+
+	_vjs.registerPlugin('apnPrebidVastPluginCommand', function(command) {
+		if (command === 'stop') {
+			if (_vastManagerObj) {
+				_vastManagerObj.stop();
+			}
+		}
+	});
+}
+
+var prebidVastPlugin = {
+	// @exclude
+	// Method exposed only for unit Testing Purpose
+	// Gets stripped off in the actual build artifact
+	test: function() {
+		return {
+			doPrebid: function(options, callback) {
+				if (window.apn_pbjs === undefined) {
+					loadPrebidScript(options, false);
+					var waitReady = setInterval(function() {
+						if (window.apn_pbjs !== undefined) {
+							clearInterval(waitReady);
+							doPrebid(options, callback);
+						}
+					}, 50);
+				}
+				else {
+					doPrebid(options, callback);
+				}
+			},
+			loadPrebidScript: loadPrebidScript,
+			apnPrebidInProgress: function() { return APN_prebid_in_progress; },
+			loadApnMolPlugin: loadApnMolPlugin,
+			renderAd: renderAd,
+			player: _player
+		};
+	},
+	// @endexclude
+
+	init: function () {
+		registerPrebidVastPlugin(videojs);
+	},
+
+	id: null,
+
+	doPrebid: function(options, id) {
+		this.id = id;
+		options.onlyPrebid = true;
+		_vjs(id).apnPrebidVastPlugin(options);
+	},
+
+	stop: function() {
+		if (this.id) {
+			_vjs(this.id).apnPrebidVastPluginCommand('stop');
+		}
+		else {
+			_vjs.getPlugins().apnPrebidVastPluginCommand('stop');
+		}
+	},
+
+	renderAd: function(renderOptions, id, creative) {
+		this.id = id;
+		renderOptions.creative = creative;
+		renderOptions.onlyPrebid = false;
+		_vjs(id).apnPrebidVastPlugin(renderOptions);
+	}
+};
+// ////////////////////////////////////////////////////////////////////
+// EXPORTS
+module.exports = prebidVastPlugin;
+
+var _player;
+var _vastManagerObj;
+var _prebidCommunicatorObj;
+if (_vjs) {
+	reqisterPrebidVastPlugin();
+}
+
+function renderAd(options) {
+	if (options.creative) {
+		// render ad if vast url is ready
+		_vastManagerObj = new _vastManager();
+		options.doPrebid = null;
+		_vastManagerObj.play(_player, options.creative, options);
+	}
+	else if (APN_prebid_in_progress) {
```

Comment:
This is a bad pattern for "waiting" for something to be done. It uses cycles for no reason every 50 milliseconds, and (depending on when the work "finishes") _wastes_ up to 50 milliseconds waiting unnecessarily. 

As long as your async methods all have callback args, there's always a way around this. You just have to define the function as "only safe to use inside the callback."

In this case it looks like you'd need to add a `callback` to `loadPrebidScript`, though, since it's the one you're waiting for.

Reply:
I'm not sure it is possible to use callback in this particular case because we call loadPrebidScript in one place but we are looking for the outcome in a different place. However, I have another solution to sync the execution of prebid during loading prebid.js and the rendering of the ad. I can use a custom event to notify the renderer that prebid has completed its work and creative is now ready for rendering. Does it work for you?

Reply:
Events would definitely better... but also introduces a lot of complexity. If you're gonna do that, I have one more suggestion, which also introduces some complexity, but offers more value too.

It looks like `loadPrebidScript` is being called in "startup" code.

Prebid.js uses a neat trick which allows publisher pages to load it `async`. It's described [here](https://github.com/prebid/Prebid.js/blob/master/src/prebid.js#L655).

Could you do something similar here, except wait to execute the queue of functions until after prebid has been loaded?

Reply:
In this case, handling the event is not so complex. We have used command queues in other products and are familiar with this solution. However, queuing commands is much more complex in this project, because the plugin can be loaded directly from the Brightcove player itself without any code on publisher page. This would make handling command queues unnecessarily difficult.

Reply:
ah! I see... I didn't know that was a workflow.

Yeah then. An event sounds good ^^